File: src/clojure/parkour/remote/sample.clj
(ns parkour.remote.sample
  (:require [parkour (conf :as conf) (wrapper :as w) (reducers :as pr)])
  (:import [java.util ArrayList Collections Random]
           [org.apache.hadoop.mapreduce InputFormat]
           [parkour.hadoop IInputFormat]
           [parkour.hadoop Sample$InputFormat]))

(defn input-format
  []
  (reify IInputFormat
    (getSplits [_ context]
      (let [klass (conf/get-class context "parkour.sample.class" nil)
            inform ^InputFormat (w/new-instance context klass)
            n (conf/get-long context "parkour.sample.n" 5)
            seed (conf/get-long context "parkour.sample.seed" 1)
            rnd (Random. seed)]
        (pr/sample-reservoir rnd n (.getSplits inform context))))
    (createRecordReader [_ split context]
      (let [klass (conf/get-class context "parkour.sample.class" nil)
            inform ^InputFormat (w/new-instance context klass)]
        (.createRecordReader inform split context)))))

File: src/java/parkour/hadoop/Sample.java
package parkour.hadoop;

import java.io.IOException;

import clojure.lang.IDeref;
import clojure.lang.IFn;
import clojure.lang.RT;
import clojure.lang.Symbol;
import clojure.lang.Var;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.mapreduce.TaskAttemptContext;

public class Sample {
  private static class Vars {
    private static final String NS = "parkour.remote.sample";
    private static final Var inputFormat = RT.var(NS, "input-format");
    static {
      RT.var("clojure.core", "require").invoke(Symbol.intern(NS));
    }
  }

  public static class InputFormat<K, V> extends ProxyInputFormat<K, V> {
    @SuppressWarnings("unchecked")
    public InputFormat() {
      super((IInputFormat) Vars.inputFormat.invoke());
    }
  }
}
